### PR TITLE
Feature/wildcard processing

### PR DESCRIPTION
**Pull request by berkayurun**
This PR improves wildcard processing performance on the Python side.

This is achieved by switching to lower cost pandas functions when possible, using a hash map instead of a list for the cache and lastly avoiding repetitive calculations by moving them outside of loops.

All changes have been tested using the sb-attacks and the example configuration found in the repository.

### CODE REVIEW

File: calculate_trigger.py [hunk 1]
```diff
@@ -228,18 +228,19 @@ def calculate_trigger_addresses(fault_list, goldenrun_tb_exec, goldenrun_tb_info
...
+                tdict is not None
+                and tdict["faultlifespan"] == fault.lifespan
+                and tdict["triggerhitcounter"] == fault.trigger.hitcounter
+                and tdict["triggeraddress"] == fault.trigger.address
+            ):
+                fault.trigger.address = tdict["answer"][0]
+                fault.trigger.hitcounter = tdict["answer"][1]
+                fault.lifespan = tdict["answer"][2]
+                found = True
+                break
```
Comment (aewag) [inline calculate_trigger.py:242]: If I am not mistaken the break statement was to break the for-loop for the cachelist.
Now the break statement breaks an "upper" for-loop.

File: calculate_trigger.py [hunk 2]
```diff
@@ -228,18 +228,19 @@ def calculate_trigger_addresses(fault_list, goldenrun_tb_exec, goldenrun_tb_info
...
+            if (
+                tdict is not None
+                and tdict["faultlifespan"] == fault.lifespan
+                and tdict["triggerhitcounter"] == fault.trigger.hitcounter
+                and tdict["triggeraddress"] == fault.trigger.address
+            ):
+                fault.trigger.address = tdict["answer"][0]
+                fault.trigger.hitcounter = tdict["answer"][1]
+                fault.lifespan = tdict["answer"][2]
+                found = True
```
Comment (aewag) [inline calculate_trigger.py:241]: Could we drop the `found` variable and instead insert a `continue` statement?

File: calculate_trigger.py [hunk 3]
```diff
@@ -133,8 +135,9 @@ def search_for_fault_location(
                     trigger_position - goldenrun_tb_info.at[idtbinfo, "ins_count"]
                 )
             else:
+                tb_id = goldenrun_tb_info.at[idtbinfo, "id"]
                 for filt in filter_lists:
-                    if filt[0] == goldenrun_tb_info.at[idtbinfo, "id"]:
+                    if filt[0] == tb_id:
                         ins = filt[len(filt) - trigger_position]
                         trigger_position = 0
                         break
```
Comment (aewag) [inline calculate_trigger.py:143]: It was not written by you, but could we change the lines to to continue if filt does not match the tb_id? For example:

```
if filt[0] != tb_id:
  continue
ins = ...
trigger_position = 0
break
```

Comment (berkayurun) [reply calculate_trigger.py:143]: Should we do the same for the lines 148 and 150?

Comment (aewag) [reply calculate_trigger.py:143]: Yes, that would be nice :)

**Review by aewag — APPROVED** [review]

File: calculate_trigger.py [hunk 4]
```diff
@@ -74,12 +74,13 @@ def calculate_lifespan_from_start(
...
-        if fault_address < tb["id"] or fault_address >= (tb["id"] + tb["size"]):
-            continue
-        tmp = goldenrun_tb_exec.index[goldenrun_tb_exec["tb"] == tb["id"]]
-        idx = idx.union(tmp)
+    matching_tbs = goldenrun_tb_info.query(
+        f"id <= {fault_address} & id + size > {fault_address}"
+    )
+    matching_tb_ids = matching_tbs["id"]  # noqa: F841
+    matching_tbs_with_positions = goldenrun_tb_exec.query("tb in @matching_tb_ids")
+    idx = pandas.Index(matching_tbs_with_positions["pos"])
```
Comment (lukasauer) [inline calculate_trigger.py:82]: The original uses the index instead of `pos` here. Is there a specific reason for using `pos`?

**Review by lukasauer — APPROVED** [review]

**Pull request merged**